src/lib/statistics/chunk_statistics/histograms/equal_width_histogram.cpp

```diff
@@ -0,0 +1,281 @@
+  Assert(bin_heights.size() == bin_distinct_counts.size(), "Must have heights and distinct counts for each bin.");
+  Assert(minimum <= maximum, "Cannot have upper bound of histogram smaller than lower bound.");
+  Assert(bin_count_with_larger_range < bin_heights.size(), "Cannot have more bins with extra value than bins.");
+  Assert(minimum.find_first_not_of(supported_characters) == std::string::npos, "Unsupported characters.");
+  Assert(maximum.find_first_not_of(supported_characters) == std::string::npos, "Unsupported characters.");
+}
+
+template <>
+EqualWidthBinData<std::string> EqualWidthHistogram<std::string>::_build_bins(
+    const std::vector<std::pair<std::string, HistogramCountType>>& value_counts, const BinID max_bin_count,
+    const std::string& supported_characters, const uint32_t string_prefix_length) {
+  // Bins shall have the same range.
+  const auto min = value_counts.front().first;
+  const auto max = value_counts.back().first;
+
+  const auto repr_min = convert_string_to_number_representation(min, supported_characters, string_prefix_length);
+  const auto repr_max = convert_string_to_number_representation(max, supported_characters, string_prefix_length);
+  const auto base_width = repr_max - repr_min + 1;
+
+  // Never have more bins than representable values.
+  const auto bin_count = max_bin_count <= base_width ? max_bin_count : base_width;
+
+  std::vector<HistogramCountType> bin_counts(bin_count);
+  std::vector<HistogramCountType> bin_distinct_counts(bin_count);
+
+  const auto bin_width = base_width / bin_count;
+  const BinID bin_count_with_larger_range = base_width % bin_count;
+
+  auto repr_current_bin_begin_value =
+      convert_string_to_number_representation(min, supported_characters, string_prefix_length);
+  auto current_bin_begin_it = value_counts.cbegin();
+
+  for (auto current_bin_id = BinID{0}; current_bin_id < bin_count; current_bin_id++) {
+    const auto repr_next_bin_begin_value =
+        repr_current_bin_begin_value + bin_width + (current_bin_id < bin_count_with_larger_range ? 1 : 0);
+    const auto current_bin_end_value = convert_number_representation_to_string(
+        repr_next_bin_begin_value - 1, supported_characters, string_prefix_length);
+
+    auto next_bin_begin_it = current_bin_begin_it;
+    while (next_bin_begin_it != value_counts.cend() && (*next_bin_begin_it).first <= current_bin_end_value) {
+      next_bin_begin_it++;
+    }
+
+    bin_counts[current_bin_id] = std::accumulate(
+        current_bin_begin_it, next_bin_begin_it, HistogramCountType{0},
+        [](HistogramCountType a, std::pair<std::string, HistogramCountType> b) { return a + b.second; });
+    bin_distinct_counts[current_bin_id] = std::distance(current_bin_begin_it, next_bin_begin_it);
+
+    current_bin_begin_it = next_bin_begin_it;
+    repr_current_bin_begin_value = repr_next_bin_begin_value;
+  }
+
+  return {min, max, bin_counts, bin_distinct_counts, bin_count_with_larger_range};
```

Comment:
afaik, if you adopt move semantic, you probably need to move all these as in `return {std::move(min), ...}`.